L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,504 @@
+struct. Similarly for `TlsServerCredentialsOptions`.
+
+### Credential Reloading
+Credential reloading will be implemented via a `CertificateProviderInterface`.
+Note that there is a naming mismatch here - there exist credentials beyond
+certificates that this _could_ theoretically support.  This provider is
+responsible for sourcing key material and supplying it to the internal stack via
+the `CertificateProvider`'s `SetKeyMaterials` API.  Two reference
+implementations of the `CertificateProviderInterface` are provided -
+`StaticDataCertificateProvider` and `FileWatcherCertificateProvider`. The
+`StaticDataCertificateProvider` provides certificates from raw string data. The
+`FileWatcherCertificateProvider` will periodically and asynchronously refresh
+certificate data from a file, allowing changes to certificates without
+restarting the process. More details about these reference implementations are
+provided below. These implementations should cover many common use cases.  If
+these providers do not support the intricacies for a specific use case, a user
+can provide their own implementations of the `CertificateProviderInterface`.
+
+```c
+/* Opaque types. */
+// A struct that stores the credential data presented to the peer in handshake
+// to show local identity. The private key and certificate chain must be PEM
+// encoded and the public key in the leaf certificate must correspond to the
+// given private key.
+struct IdentityKeyCertPair {
+  std::string private_key_pem;
+  std::string certificate_chain_pem;
+};
+
+
+/** ------------------------------------ Provider ------------------------------------ */
+/** Provides identity credentials and root certificates.
+/* To enable the provider to manage multiple sets of credentials of each type,
+ * each set of credentials must be given a "name". The "name" is an opaque
+ * user-defined label, and the provider will reference this "name" e.g. in logs
+ * and error messages.
+ */
+class CertificateProviderInterface {
+ public:
+  virtual ~CertificateProviderInterface() = default;
+
+  // Provider implementations MUST provide a WatchStatusCallback that will be
+  // called by the internal stack. This will be invoked when a new certificate
+  // name is starting to be used internally, or when a certificate name is no
+  // longer being used internally. When being invoked for a new certificate,
+  // this callback should call `SetKeyMaterials` to do the initial population
+  // the certificate data in the internal stack.
+  // 
+  // For the parameters in the callback function:
+  // cert_name The name of the certificates being watched.
+  // root_being_watched If the root certificates with the specific name are being
+  // watched.
+  // identity_being-Watched If the identity certificates with the specific name
+  // are being watched.
+  virtual void WatchStatusCallback(std::string cert_name, bool root_being_watched, bool identity_being_watched) = 0;
```

Comment:
From review, need to rename this method.

Reply:
Discussed elsewhere, I believe we decided `OnWatchStarted` and `OnWatchStopped`.
Going to resolve this in an effort to clean up.